### ISSUE TITLE
[CT-3478] [Bug] Editing profiles as part of tests is not thread-safe

### ISSUE THREAD

**Issue by benc-db** [dbt-labs/dbt-core#9254]
### Is this a new bug in dbt-core?

- [X] I believe this is a new bug in dbt-core
- [X] I have searched the existing issues, and I could not find an existing issue for this bug

### Current Behavior

When running tests in parallel that modify the profile, it is possible for invocations of dbt_run to grab the wrong profile, even across different test classes.

In dbt/tests/fixtures/project.py, you have:

```
@pytest.fixture(scope="class")
def profiles_yml(profiles_root, dbt_profile_data):
    os.environ["DBT_PROFILES_DIR"] = str(profiles_root) # <- this is not safe, as it is global
    write_file(yaml.safe_dump(dbt_profile_data), profiles_root, "profiles.yml")
    yield dbt_profile_data
    del os.environ["DBT_PROFILES_DIR"] # <- this is not safe, as it is global
```

### Expected Behavior

When you modify a profile as part of a test fixture, the modified profile is read as part of the test.

### Steps To Reproduce

This is hard to reproduce consistently, but basically:
1. Have multiple tests that run at the same time that modify/rely on a config in the profile
2. Run them in parallel
3. Observe sporadic failure

### Environment

```markdown
- OS: Mac
- Python: Multiple
- dbt: 1.7.3
```

### Which database adapter are you using with dbt?

other (mention it in "Additional Context")

### Additional Context

Trying to write tests for dbt-databricks for warehouse-per-model.

**Comment by jtcohen6** [comment]
@benc-db Fair point! Rather than resetting an environment variable (that sets a flag pointing to a file path), it would be better if the profile + project data could be passed directly into the `dbtRunner` programmatically [here](https://github.com/dbt-labs/dbt-core/blob/1740df534bae3e0a639eb97df17bbb8f6ba4e704/core/dbt/tests/util.py#L92-L100).

In the meantime - rather than modifying the profile data within your tests, I'd recommend setting up the test case with multiple profiles, and then rerouting via the `--target` flag. Here's an example where we do this:

https://github.com/dbt-labs/dbt-core/blob/1740df534bae3e0a639eb97df17bbb8f6ba4e704/tests/functional/run_operations/test_run_operations.py#L28-L57

https://github.com/dbt-labs/dbt-core/blob/1740df534bae3e0a639eb97df17bbb8f6ba4e704/tests/functional/run_operations/test_run_operations.py#L85-L86
